File: libtorch/src/satadj.cpp
#include "kdm6/satadj.h"

namespace kdm6 {
namespace satadj {

SatAdjParams default_satadj_params(double rv) {
    return SatAdjParams{
        /*rv=*/rv,
        /*qmin=*/constants::EPS,      // Fortran qmin=epsilon=1e-15 (driver QMIN). satadj uses it as the
                                      // q_eff=max(q,qmin) floor (F:2927) — same #1 issue, same fix as ThermoParams.
                                      // A pure floor (no div-safety), inactive in warm cells where q>>floor.
    };
}

SatAdjOutputs saturation_adjustment_torch(
    const torch::Tensor& t,
    const torch::Tensor& q,
    const torch::Tensor& qc,
    const torch::Tensor& qs1,
    const torch::Tensor& xl,
    const torch::Tensor& cpm,
    const SatAdjParams& params,
    double dtcld
) {
    // work1 = (max(q, qmin) - qs1) / (1 + xl² · qs1 / (rv · cpm · t²))
    auto t_safe = torch::clamp(t, /*min=*/1.0);
    // Fortran conden stmt-fn (module_mp_kdm6.F:781, used at F:2927):
    //   denom = 1.+ d*d/(rv*e)*c/(a*a)   with d=xl, e=cpm, c=qs1, a=t.
    // Fortran `*` and `/` are equal-precedence, left-associative, so the term is
    //   ((((xl*xl)/(rv*cpm))*qs1)/(t*t)) — divide by (rv*cpm) FIRST, then *qs1,
    // then /(t*t). gfortran (-O2, no -ffast-math: configure.wrf:150) does NOT
    // reassociate division, so this left-to-right grouping is what mp37 generates.
    // Mirror that grouping here (NOT (xl*xl*qs1)/(rv*cpm*t*t)) so the float32
    // operational path bit-matches mp37 in this saturation-adjustment kernel — the
    // condensate-onset seed the prior FMA/libm sweep could not reach (division
    // reassociation is not an FMA contraction). Tensor-ops only ⇒ autograd-safe.
    // (f64-activation experiment 2026-06-10: upcasting this conden to f64 left the
    // QCLOUD divergence vs mp37 BYTE-UNCHANGED — the residual is NOT the conden
    // cancellation floor; it is already present in the t/q/qs1 ENTERING satadj, i.e.
    // upstream. Reverted to native f32.)
    auto denom = 1.0 + xl * xl / (params.rv * cpm) * qs1 / (t_safe * t_safe);
    auto qmin_t = torch::full_like(q, params.qmin);
    auto q_eff = torch::maximum(q, qmin_t);
    auto work1 = (q_eff - qs1) / denom;

    // Branch on sign:
    //   work1 > 0          → cond_path = min(work1, max(q, 0)) / dtcld
    //   work1 < 0 and qc>0 → evap_path = max(work1, -qc) / dtcld
    //   else               → 0
    auto zero = torch::zeros_like(q);
    // §44 DOUBLE-internal-narrowed-once: Fortran work1 is DOUBLE (F:660), dtcld REAL(4) →
    // work1/dtcld + the min/max run in DOUBLE, narrowing to the REAL(4) pcond ONCE at the
    // store (F:3020/3022). C++ ran the whole chain in f32, rounding at every op. Mirror Fortran:
    // promote work1/q/qc to f64, do /dtcld + min/max in f64, narrow at op dtype (q.scalar_type():
    // f32 operational = Fortran-faithful; f64 DA = no-op, VJP preserved). work1 itself stays
    // native f32 (F:3018 conden is REAL(4); 2026-06-10 f64-conden experiment was byte-neutral).
    // The complete-evap RHS below stays PURE f32 (-qc/dtcld): Fortran F:3027 compares the
    // f32-narrowed pcond against an f32 -qci/dtcld (qci,dtcld both REAL(4)) — do NOT upcast it.
    auto rdt = q.scalar_type();
    auto wd = work1.to(torch::kFloat64);
    auto qd = q.to(torch::kFloat64);
    auto qcd = qc.to(torch::kFloat64);
    auto cond_path = (torch::minimum(wd, torch::maximum(qd, torch::zeros_like(qd))) / dtcld).to(rdt);
    auto evap_path = (torch::maximum(wd, -qcd) / dtcld).to(rdt);

    auto is_super = work1 > 0;
    auto is_sub_with_cloud = torch::logical_and(work1 < 0, qc > 0);

    auto pcond = torch::where(
        is_super, cond_path,
        torch::where(is_sub_with_cloud, evap_path, zero)
    );
    // Fortran F:2966 is a BARE equality gate `if(pcond.eq.-qci(i,k,1)/dtcld)` — no
    // qc>0 / sub-saturation conjunct. At an entry-clamped qc==0 cell with pcond==0
    // it fires (0.0 == -0.0), transferring nc -> nccn (the step-46 nn seed when the
    // old is_sub_with_cloud gate blocked it). pcond>=0 with qc>0 makes -qc/dt<0
    // unreachable, so the bare gate is exactly the evap-exhausted + qc==0 set.
    auto cloud_complete_evap = (pcond == (-qc / dtcld));
    return SatAdjOutputs{/*pcond=*/pcond, /*cloud_complete_evap=*/cloud_complete_evap};
}

}  // namespace satadj
}  // namespace kdm6

File: oracle/kdm6/satadj.py
"""
KDM6 saturation adjustment — Step B5.

원본: module_mp_kdm6.F: 2927-2943

Microphysics 끝에서 qv ↔ qc 1-step Clausius-Clapeyron 보정. latent heat
feedback (`xl²/(rv·cpm)·qs/t²` 항)으로 응결/증발이 t/q를 동시에 변경하는 effect를
1차 근사한다.

Algorithm (Fortran 직역):
    work1 = (max(q, qmin) - qs) / (1 + xl²/(rv·cpm) · qs/t²)
    if work1 > 0:        # supersaturation
        pcond = min(work1/dtcld, q/dtcld)        # cap by available qv
    elif qc > 0 and work1 < 0:   # subsaturation with available qc
        pcond = max(work1, -qc) / dtcld          # cap by available qc
    else:
        pcond = 0

Notes
-----
- nccond (cloud number transfer to CCN at complete evap)는 `state mutation`이므로
  rain_evap와 동일하게 caller가 처리. oracle은 pcond rate만 산출.
- saturation vapor pressure (qs1 자체) 계산은 별도 sub-step 또는 외부 진단.
  본 함수는 qs1을 placeholder input으로 받음.
- `conden`의 분모 `1 + xl²·qs/(rv·cpm·t²)`는 항상 > 1이라 division-by-zero 위험 없음.
"""
from __future__ import annotations

from typing import NamedTuple

import torch

from . import constants as c


# ─── physical constants (default values, kdm6init INPUT) ─────────────────────

DEFAULT_RV = 461.6  # J/kg/K, Fortran r_v (module_model_constants.F:22)


class SatAdjParams(NamedTuple):
    """saturation adjustment에 필요한 thermodynamic constant."""

    rv: float
    qmin: float


def default_satadj_params(*, rv: float = DEFAULT_RV) -> SatAdjParams:
    return SatAdjParams(rv=rv, qmin=c.EPS)  # Fortran qmin=epsilon=1e-15; q_eff=max(q,qmin) floor (F:2927). 1:1 fix #1 (satadj path).


def saturation_adjustment_torch(
    t: torch.Tensor,
    q: torch.Tensor,
    qc: torch.Tensor,
    qs1: torch.Tensor,    # saturation mixing ratio w.r.t. water (외부 진단)
    xl: torch.Tensor,     # latent heat of vaporization [J/kg] (외부 진단)
    cpm: torch.Tensor,    # moist heat capacity [J/kg/K] (외부 진단)
    *,
    params: SatAdjParams,
    dtcld: float,
) -> torch.Tensor:
    """Fortran 2927-2931 — saturation adjustment rate.

    Returns
    -------
    pcond : (B, K) tensor [kg/kg/s]. > 0이면 응결 (qv→qc), < 0이면 증발 (qc→qv).
    """
    qmin_t = torch.tensor(params.qmin, dtype=q.dtype, device=q.device)

    # work1 = (max(q, qmin) - qs1) / (1 + xl²·qs1/(rv·cpm·t²))   — Clausius-Clapeyron
    # 분모는 항상 > 1 (양수 보장). 단, t² 자체는 분모이므로 0 회피 위해 clamp.
    t_safe = torch.clamp(t, min=1.0)  # T > 1 K 가정 (대기 환경 100~400 K)
    # Fortran conden stmt-fn (module_mp_kdm6.F:781): denom = 1.+ d*d/(rv*e)*c/(a*a),
    # equal-precedence left-assoc => ((((xl*xl)/(rv*cpm))*qs1)/(t*t)). gfortran does
    # not reassociate division, so mirror that grouping (NOT (xl²·qs1)/(rv·cpm·t²))
    # to stay 1:1 with F:781 and with the C++ port (satadj.cpp:27).
    denom = 1.0 + xl * xl / (params.rv * cpm) * qs1 / (t_safe * t_safe)
    q_eff = torch.maximum(q, qmin_t)
    work1 = (q_eff - qs1) / denom

    # Branch on sign:
    #   work1 > 0  → super-saturated, condensation: pcond = min(work1/dtcld, q/dtcld)
    #   work1 < 0 AND qc > 0 → sub-saturated with available cloud: pcond = max(work1, -qc)/dtcld
    #   otherwise  → 0
    cond_path = torch.minimum(work1, torch.maximum(q, torch.zeros_like(q))) / dtcld
    evap_path = torch.maximum(work1, -qc) / dtcld

    is_super = work1 > 0
    is_sub_with_cloud = (work1 < 0) & (qc > 0)

    zero = torch.zeros_like(q)
    pcond = torch.where(
        is_super, cond_path,
        torch.where(is_sub_with_cloud, evap_path, zero)
    )
    return pcond


__all__ = [
    "SatAdjParams",
    "default_satadj_params",
    "saturation_adjustment_torch",
]
